app/models/miq_ae_namespace.rb

```diff
@@ -8,25 +11,39 @@ class MiqAeNamespace < ApplicationRecord
                          /^commit_time/, /^commit_sha/, /^ref$/, /^ref_type$/,
                          /^last_import_on/, /^source/, /^top_level_namespace/].freeze
 
-  belongs_to :parent,        :class_name => "MiqAeNamespace",  :foreign_key => :parent_id
-  has_many   :ae_namespaces, :class_name => "MiqAeNamespace",  :foreign_key => :parent_id,    :dependent => :destroy
-  has_many   :ae_classes, -> { includes([:ae_methods, :ae_fields, :ae_instances]) },    :class_name => "MiqAeClass",      :foreign_key => :namespace_id, :dependent => :destroy
+  has_many :ae_classes, -> { includes([:ae_methods, :ae_fields, :ae_instances]) }, :class_name => "MiqAeClass", :foreign_key => :namespace_id, :dependent => :destroy
+
+  validates_format_of :name, :with    => /\A[\w\.\-\$]+\z/i,
+                             :message => N_("may contain only alphanumeric and _ . - $ characters")
+  validates :name, :presence => true, :uniqueness => {:scope => :ancestry, :case_sensitive => false}
+
+  alias ae_namespaces children
+  virtual_has_many :ae_namespaces
 
-  validates_presence_of   :name
-  validates_format_of     :name, :with    => /\A[\w\.\-\$]+\z/i,
-                                 :message => N_("may contain only alphanumeric and _ . - $ characters")
-  validates_uniqueness_of :name, :scope => :parent_id, :case_sensitive => false
+  after_save :clear_lookup_by_fqname_cache
 
   def self.lookup_by_fqname(fqname, include_classes = true)
     return nil if fqname.blank?
 
-    fqname   = fqname[0] == '/' ? fqname : "/#{fqname}"
-    fqname   = fqname.downcase
-    last     = fqname.split('/').last
+    @lookup_by_fqname_cache ||= {}
+    fqname = fqname[1..-1] if fqname[0] == '/'
+    fqname = fqname.downcase
+    return @lookup_by_fqname_cache[fqname] if @lookup_by_fqname_cache.key?(fqname)
+
+    domain, *partial = fqname.split('/')
     low_name = arel_table[:name].lower
-    query    = includes(:parent)
-    query    = query.includes(:ae_classes) if include_classes
-    query.where(low_name.eq(last)).detect { |namespace| namespace.fqname.downcase == fqname }
+    query = include_classes ? includes(:ae_classes) : all
+
+    @lookup_by_fqname_cache[domain] ||= query.roots.find_by(low_name.eq(domain))
+    parent_name = domain
+    partial.each do |ns|
+      key_name = "#{parent_name}/#{ns}"
+      @lookup_by_fqname_cache[key_name] ||= @lookup_by_fqname_cache[parent_name]&.children&.find_by(low_name.eq(ns))
```

Comment:
The purpose of ancestry is to not have multiple lookups (i.e. `find_by`)

If we need to do this, lets store the fully qualified path and use that to look it up:

```ruby
self.ancestry_name = path.map(&:name).map(&:downcase).join("/")

where(:ancestry_name => fqname.downcase)
```

Reply:
moved to future PR